restore socket descriptor permission management

tyrantlizard77 opened the pull request:
A change made to the socket system calls to use actual file descriptors (instead of `struct net_context` stuffed into an integer) unintentionally lost the ability to manage per-thread permissions of file descriptors, which is needed in MPU-based user mode memory models (currently all arches that implement user mode in Zephyr).

This PR:

- Adds an API `z_is_in_user_syscall()` so that system call verification checks may be done deep within implementation functions if those checks are only relevant to syscalls originating from user mode.
- All network context object types which have `__net_socket` in their declaration will have their instances tracked as kernel objects with type `K_OBJ_NET_SOCKET`
- Socket APIs once again check for object permissions after the object is retrieved from file descriptor tables. This check only applies if the system call originated from user mode.
- Add a custom API that, given a file descriptor, returns a pointer to the associated `struct net_context` so that APIs like `k_object_access_grant()` may be used.
- Add tests for new APIs, and show that permission management is working as expected across user threads.
- Add some infrastructure for registering `K_OBJ_NET_SOCKET` objects dynamically, needed by the `socketpair` implementation (which pulls data from a heap instead of a pool like other socket types).

Please review individual commit messages for more information.

Process-based virtual memory is coming to Zephyr for systems with an MMU later this year. In the fullness of time, we can adopt file descriptors that have process-level scope (instead of global scope), removing the need for this management on those types of systems. We will likely continue to need this infrastructure indefinitely for MPU-based systems, but it is the same permission model we already use for all kernel objects and device driver instances.

Commit:
userspace: add z_is_in_user_syscall()

Certain types of system call validation may need to be pushed
deeper in the implementation and not performed in the verification
function. If such checks are only pertinent when the caller was
from user mode, we need an API to detect this situation.

This is implemented by having thread->syscall_frame be non-NULL
only while a user system call is in progress. The template for the
system call marshalling functions is changed to clear this value
on exit.

A test is added to prove that this works.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: check permissions on net contexts

The original sockets system calls used file descriptors which
were actually net_context pointers. For all socket system calls,
any calls from user mode would check if the caller had permission
to use the net context.

This was later changed to not stuff net_context pointers into file
descriptors, but all the permission checking was unintentionally
lost, allowing all threads on the system to
read/write all socket file descriptors in the system at will, with
no way to isolate applications running on the same microcontroller
from each other's network activity.

This patch restores the permission checks on network context objects
for socket system calls that originated from user mode.

The call to z_object_recycle() was never removed from
zsock_socket_internal(); this is again leveraged to grant the
caller who opened the socket permission on the net_context
associated with the returned file descriptor.

To ensure that all socket calls do this checking, all uses of
z_get_fd_obj_and_vtable() have been routed through get_sock_vtable().

Objects have initialization state set and thread permissions
reset to just the caller in common zsock_socket() code.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: sockets: add API to fetch an fd's ctx object

Zephyr running on MPU devices have a different memory model than
process-oriented OSes like Linux and require a method to set
kernel object permissions on a file descriptor's underlying
context object. Add this, and a test to show that it is working.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: fix bad ssf pointer on bad/no syscall

This was passing along _current->ssf, but these types of bad
syscalls do not go through the z_mrsh mechanism and was
passing stale data.

We have the syscall stack frame already as an argument,
propagate that so it works properly.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
@jukkar I have this almost working.

It works as expected for regular native sockets, as they are all backed by a `struct net_context` which I've restored as a kernel object to track permissions.

It doesn't work for registered socket families with `NET_SOCKET_REGISTER()`, as these are not backed by a `struct net_context` that I can see, instead custom data structures which are dependent on the implementation. I didn't see this until I started trying to run tests that use them like tests/net/socket/net_mgmt.

I've just now started looking at this part and will hopefully come up with a generic solution.

andrewboie commented:
@jukkar think I found something workable.

There's no common object type I can use for underlying socket descriptor objects, so it's not as simple as the last time I looked in this code, where you could make `net_context` a kernel object and call it a day.

Any given socket registered with NET_SOCKET_REGISTER() will have their own object type that is returned when you call z_get_fd_obj_and_vtable(). It's an arbitrary void * type.

However, I can use a technique that was contributed to make driver disambiguation possible, because in order to distinguish drivers, we needed to be able to identify what subsystem it belongs to. If you look through subsystem headers you'll see that the API struct for all subsystems is tagged with `__subsystem` in order to identify it.

I plan on doing something similar to tag socket descriptor object types with `__net_socket_object` or something similar. Things like `struct modem_socket`, `struct net_mgmt_socket`, etc, basically any struct which represents a socket fd object would be tagged with this.

Then the kernel object tables will have entries for any struct instance it finds that was tagged with `__net_socket_object`, and they will all have the same kernel object type that we can check. The fact that they all have the same object type signifies that they are compatible with methods in `struct socket_op_vtable` and are (from the core socket code's perspective) interchangeable.

We'll be able to handle any kind of socket object then.

andrewboie commented on include/net/socket.h:
```diff
@@ -145,6 +145,39 @@ struct zsock_addrinfo {
+ * @return pointer to associated struct net_context, or NULL if the
+ *         file descriptor wasn't valid or the caller had no permission
+ */
+__syscall struct net_context *zsock_get_context_object(int sock);
```
this is wrong, socket objects are `void *`, I got confused from older code where everything was backed by a `net_context` but it's much more abstracted now. Will fix this and other changes that coerced fd object types into a `net_context` pointer, needs to be left as `void *`.

jukkar commented on tests/net/socket/tcp/src/main.c:
```diff
@@ -446,8 +446,73 @@ void test_v4_accept_timeout(void)
 	k_sleep(TCP_TEARDOWN_TIMEOUT);
 }
 
+#ifdef CONFIG_USERSPACE
```
The tests look good. I wish we had these originally so would not have missed the issue.
We probably also need to add tests to other socket types like net_mgmt, websocket etc.

andrewboie replied:
Agreed, I think the end goal we should strive for is to exercise the cross product of all socket APIs against all socket types.

jukkar commented:
> I want to try to do something similar to tag socket descriptor object type struct definitions with `__net_socket_object` or something similar. Things like `struct modem_socket`, `struct net_mgmt_socket`, etc, basically any struct which represents a socket fd object would be tagged with this.

Sounds reasonable. So if I understood you right, we would add a `__net_socket_object` tag to those structs that can be "linked" to a socket descriptor. And then do some voodoo magic in a script that would generate something from "that information". Atm I am not very familiar what that "information" is, but I suppose I will learn soon :)

Commit:
scripts: gen_kobject_list: generalize obj alloc

Instead of handling this ad hoc, generalize which kobject
types can be allocated.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
scripts: parse_syscalls: generalize struct tags

Now we can build up lists of data structures matching a list
of particular tags, with __subsystem being just one case.

Relax searches to also look inside C files, since struct
prototypes may be declared there as well.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: net sockets are kernel objects

Any data structure declaration tagged with __net_socket will end up
in the kernel object table with type K_OBJ_NET_SOCKET. These all
correspond to objects which are associated with socket file
descriptors and can handle the socket vtable API.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: tag net socket objects

Used for permission validation when accessing the associated file
descriptors from user mode.

There often get defined in implementation code, expand the search
to look in drivers/ and subsys/net/.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: net_mgmt: grant socket access

This socket is shared by all the test cases which run in
different threads. Just make it a global object here.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
tests: net: tcp: test zsock_get_context_object

Add test case to prove that this new API works.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: lib: remove socket-specific recycle calls

This is just done in common code now.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
fdtable: init fd context objects

Anytime a file descriptor context object is updated, we need to
reset its access permissions and initialization state. This
is the most centralized place to do it.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
userspace: improve dynamic object allocation

We now have a low-level function z_dynamic_object_create()
which is not a system call and is used for installing
kernel objects that are not supported by k_object_alloc().

Checking for valid object type enumeration values moved
completely to the implementation function.

A few debug messages and comments were improved.

Futexes and sys_mutexes are now properly excluded from
dynamic generation.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
net: socketpair: fix user mode access

The socketpair file descriptor context objects are heap allocated
and not drawn from a static pool. Register these as kernel objects
when we create them if user mode is enabled.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

Commit:
kernel: fix k_object_free() spelling

This was supposed to match definitions if dynamic objects
are turned on.

Signed-off-by: Andrew Boie <andrew.p.boie@intel.com>

andrewboie commented:
> Sounds reasonable. So if I understood you right, we would add a `__net_socket_object` tag to those structs that can be "linked" to a socket descriptor. And then do some voodoo magic in a script that would generate something from that "information". Atm I am not very familiar what that "information" is, but I suppose I will learn soon :)

Yeah, there's some logic that scans code looking for these special function tags like `__syscall` and struct declaration tags like `__subsystem`. I augmented it to also look for `__net_context` and set up kernel object metadata appropriately.

andrewboie commented:
@jukkar @tbursztyka this is passing all my tests now. But I'm just running `sanitycheck`, I know there are additional scenarios. Review comments and any additional testing you could provide would be gratefully appreciated.

I implemented the `__net_socket` stuff I alluded to earlier, and also set up the dynamic kernel object installation needed by the socketpair code. Do we know why socketpair uses k_malloc for context objects instead of a pool like the other socket types?

The only remaining thing were the socket objects for two wifi drivers: simplelink and esp32. Problem is, the object for these descriptors isn't a location in memory but an integer stuffed into a `void *`. Not sure why it's like that. At any rate esp32 doesn't even support user mode (it's xtensa) I'd like to just leave this alone for now.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -45,8 +45,36 @@ const struct socket_op_vtable sock_fd_op_vtable;
+				      (const struct fd_op_vtable **)vtable);
+
+#ifdef CONFIG_USERSPACE
+	if (z_is_in_user_syscall()) {
```
Should we check here that ctx is not null, so `if (ctx && z_is_in_user_syscall()) {`

andrewboie replied:
z_object_find() returns NULL if you pass it NULL (or any address that isn't a kernel object). assuming the success case is far more likely, I think it's slightly more efficient to omit a check at this function's level.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -45,8 +45,36 @@ const struct socket_op_vtable sock_fd_op_vtable;
+			z_dump_object_error(ret, ctx, zo, K_OBJ_NET_SOCKET);
+			/* Invalidate the context, the caller doesn't have
+			 * sufficient permission or there was some other
+			 * problem with the net context object
```
Probably better to say here `problem with the net socket object`

andrewboie replied:
done.

jukkar commented on subsys/net/lib/sockets/sockets.c:
```diff
@@ -45,8 +45,36 @@ const struct socket_op_vtable sock_fd_op_vtable;
+#endif /* CONFIG_USERSPACE */
+
+	if (ctx == NULL) {
+		NET_ERR("invalid access on fd %d by thread %p", sock,
```
Instead of talking about fd, better say `Invalid access on sock %d by thread %p` as it is not a file descriptor but socket one.

andrewboie replied:
done.

jukkar reviewed (commented):
> But I'm just running `sanitycheck`, I know there are additional scenarios. Review comments and any additional testing you could provide would be gratefully appreciated.

I will ask @rmstoi to run Maxwell tests for this PR. I will also do manual testing. Unfortunately most of our testing apps are not using usermode, I will investigate this more if we could add more usermode applications.

andrewboie commented:
@jukkar thanks for reviewing, updated based on your feedback.

jukkar commented:
> I will ask @rmstoi to run Maxwell tests for this PR.

Maxwell did not reveal any issues. I think this PR is ready for merge.

jukkar commented:
@andrewboie was there an issue for this PR?

ceolin commented on include/kernel.h:
```diff
@@ -352,6 +352,27 @@ void k_object_access_all_grant(void *object);
+ * Initialization state and thread permissions will be cleared. The
+ * returned z_object's data value will be uninitialized.
+ *
+ * Most users will want to use k_object_alloc() instead.
```
stupid question, what are the objects not supported by `k_object_alloc` ?

andrewboie replied:
1. futexes and sys_mutex, they live in user and not kernel memory
2. thread stacks, as we do not have an aligned allocator
3. net sockets (this patch) as they are a family of objects of no specific size

There is a switch statement in `z_impl_k_object_alloc()` specifically for this

andrewboie replied:
This could be better documented by the way, I'll send a separate PR to augment the doxygen.

ceolin replied:
thanks !

pfalcon commented:
> This PR:

>    Adds an API z_is_in_user_syscall() so that system call verification checks may be done deep within implementation functions if those checks are only relevant to syscalls originating from user mode.

I like this specific point, because it resonates well with a response I have in mind regarding matters brought in the other ticket: All functionality of "emulation" APIs, like e.g. POSIX, is implemented in terms of native Zephyr APIs. And if native Zephyr has a requirement and desire to add overheads of checks, then that's exactly what should happen. It shouldn't be the case that all checks are tried to performed on the entry to kernel, and afterwards, the control flow turns into amorphous, onwer-less, permission-less, Brownian motion.

Instead, there should be thread-specific execution (and thus, permission) context associated with a thread even after entering kernel, and all relevant native API functions in kernel should perform permission checks as they get called. And the idea to "snapshot the world" and perform all possible checks on entering the kernel seems rather cumbersome and rather flaky in the first place.

> if those checks are only relevant to syscalls originating from user mode.

In this model, all syscalls ultimately originate in user mode. Kernel just doesn't open sockets or files for itself. And desire to let some *ultimately user* threads to run in kernel mode and bypass security checks just because of that seems like another flaky idea.  As long as all these checks can be configured out and compiled out (i.e. don't have kernel/userspace split on the first place), going out of the way to add backdoors to bypass checks inside the kernel functions doesn't seem to worth the trouble it takes and may cause if something's overlooked somewhere.

ioannisg commented on include/syscall_handler.h:
```diff
@@ -25,6 +25,38 @@ enum _obj_init_check {
+ * @return whether the current context is handling a syscall for a user
+ *         mode thread
+ */
+static inline bool z_is_in_user_syscall(void)
```
Can this be called from ISR @andrewboie ?

andrewboie replied:
No it cannot, good catch. I'll add an assertion.

ioannisg replied:
Should be ok now.
You could optionally add a comment in the description that the function should return false if called in isr context

andrewboie commented:
> Instead, there should be thread-specific execution (and thus, permission) context associated with a thread even after entering kernel, and all relevant native API functions in kernel should perform permission checks as they get called. And the idea to "snapshot the world" and perform all possible checks on entering the kernel seems rather cumbersome and rather flaky in the first place.

It's really quite breathtaking how forcefully you weigh your opinions on topics that you are so obviously uninformed on. Go read about security features like SMEP/SMAP, or contemplate the fundamental differences and constraints on handling between user vs kernel pointers if you don't understand why we are "snapshotting the world" upon entry/exit to system calls.

While you're doing that, given that you're the one who broke this to begin with, maybe you could lay off the trolling at least in this PR?

pfalcon commented:
@andrewboie 

> handling between user vs kernel pointers

My questions in https://github.com/zephyrproject-rtos/zephyr/pull/25443#issuecomment-634518560 were exactly about details of validating user vs kernel pointers. And you refused (or were unable) to answer them.

> While you're doing that, given that you're the one who broke this to begin with, maybe you could lay off the trolling at least in this PR?

I'm trying to share my load of maintenance of this stuff, and learn as I go. If you're not interested, or unable, to cooperate, and discuss various aspects of this matter, then feel to maintain and update it yourself as various subsystems evolve (and apparently accuse people of breaking it).

carlescufi commented:
Merging this now so it makes 2.3.0-rc2

pfalcon commented:
@carlescufi, @galak: FYI, this failed build in our CI for gnuarmemb toolchain (but not Zephyr SDK). The overall picture: https://ci.linaro.org/job/zephyr-upstream/5245/ . Example of failed build: https://ci.linaro.org/job/zephyr-upstream/5245/PLATFORM=frdm_k64f,ZEPHYR_TOOLCHAIN_VARIANT=gnuarmemb,label=docker-xenial-amd64-13/console . The error:

~~~
/home/buildslave/workspace/zephyr-upstream/zephyr/kernel/userspace.c: In function 'z_impl_k_object_alloc':
/home/buildslave/workspace/zephyr-upstream/zephyr/kernel/userspace.c:319:22: error: 'tidx' may be used uninitialized in this function [-Werror=maybe-uninitialized]
   zo->data.thread_id = tidx;
   ~~~~~~~~~~~~~~~~~~~^~~~~~
/home/buildslave/workspace/zephyr-upstream/zephyr/kernel/userspace.c:285:12: note: 'tidx' was declared here
  uintptr_t tidx;
            ^~~~
cc1: all warnings being treated as errors
~~~
